### PR TITLE
Reorganize and expand the testing chapters.

### PR DESCRIPTION
**Pull request by ehuss**
This is an attempt to reorganize the layout of the testing chapters into something that makes a little more sense to me (though probably not everyone).
The new chapter layout is:

- Testing the compiler — Introduction and high-level overview of different kinds of tests.
    - Running tests — How to run tests.
        - Testing with Docker — How to run tests with Docker.
        - Testing with CI — How to run tests with GitHub Actions.
    - Adding new tests — A basic overview and tutorial of adding a test.
    - Compiletest — Reference on how compiletest works and the different test suites.
        - UI tests — Reference on writing UI tests.
        - Test headers — Reference on test headers.
    - Performance testing — Intro to rustc-perf testing.
    - Crater — Overview of crater testing.

This also includes a lot of new content to fill out these sections.

### CODE REVIEW

**Comment by ehuss** [comment]
cc @petrochenkov who mentioned they may be willing to look this over.

**Comment by ehuss** [comment]
Thanks for taking a look!  I have applied some fixes.

File: src/tests/ui.md [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+  - Example: `/path/to/rust/build/x86_64-unknown-linux-gnu/test/ui`
+
+Additionally, the following changes are made:
+
+- Line and column numbers for paths in `$SRC_DIR` are replaced with `LL:CC`.
+  For example, `/path/to/rust/library/core/src/clone.rs:122:8` is replaced with
+  `$SRC_DIR/core/src/clone.rs:LL:COL`.
+
+  Note: The line and column numbers for `-->` lines pointing to the test are
+  *not* normalized, and left as-is. This ensures that the compiler continues
```
[the remark this answers is not in the document]
Comment (ehuss) [reply src/tests/ui.md]: I rewrote this section a little, as it was also a little outdated and wasn't quite as clear as I'd like.

**Comment by camelid** [comment]
@ehuss could you mark the accepted fixes as resolved to make it easier to follow along with what's been changed? Thanks!

**Review by spastorino — APPROVED** [review]

**Comment by spastorino** [comment]
This is great, there are after some other merges some conflicts. Can you rebase?.

**Comment by ehuss** [comment]
Thanks, I have rebased.

**Pull request merged**